### PR TITLE
All / none button for multi select component

### PR DESCRIPTION
**Pull request by rubenmoya**
This PR implements #9502

allnone

### Acceptance
- [x] ALL selects all the items
- [x] NONE deselects all the items
- [x] ALL and NONE are gone if query search is applied

### CODE REVIEW

**Comment by xavijam** [comment]
Already mentioned off-line:

- Buttons should be smaller.
- Paddings are not correct.
- Search should be bigger, we need to gain more space.

**Comment by rubenmoya** [comment]
Is this better? I'm using `u-lSpace--xl`, `u-lSpace` is too small imho, and there is no `u-lSpace--m`

screen shot 2017-05-25 at 17 49 07

**Comment by xavijam** [comment]
About space and padding and so on, perfect, but buttons are still big.

**Comment by rubenmoya** [comment]
That's true, this looks better :)

screen shot 2017-05-25 at 17 56 47

**Comment by xavijam** [comment]
NOW WE ARE TALKING MY FRIEND! 👍

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-search-view.js [hunk 1]
```diff
@@ -36,7 +36,12 @@ module.exports = CoreView.extend({
 
   _checkButtons: function () {
     var query = this.model.get('query');
-    this.$('.js-clear').toggle(query.length > 0);
+
+    if (query.length > 0) {
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-search-view.js]: What about?

```js
this.$('.js-clear').css('opacity', query.length ? 0 : 1);
```

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-search.tpl [hunk 2]
```diff
@@ -1,6 +1,6 @@
-<div class="CDB-Box-modalHeaderItem">
+<div class="u-flex">
   <input type="text" name="text" autocomplete="off" placeholder="<%- searchPlaceholder %>" class="CDB-InputTextPlain CDB-Text js-search">
-  <button type="button" style="display:none" class="u-lSpace--xl js-clear">
+  <button type="button" style="opacity: 0;" class="u-lSpace--xl js-clear">
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-search.tpl]: Could we create a class and put it wherever we consider? We should not add inline styles. Let's do it! 💃

File: app/assets/stylesheets/editor-3/custom-list.css.scss [hunk 3]
```diff
@@ -119,3 +119,7 @@
 .CustomList-list.ps-container > .ps-scrollbar-x-rail {
   margin-bottom: 2px;
 }
+
+.CustomList-search {
```
Comment (piensaenpixel) [inline app/assets/stylesheets/editor-3/custom-list.css.scss]: IMO this should a modifier :p and like we have `.CustomList--small` etc... what do you think if we use something like `.CustomList--full`
BTW we have `.CustomList-itemLink` and maybe we could consolidate this one.
My 2 ctmos.

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-view.js [hunk 4]
```diff
@@ -84,17 +95,29 @@ module.exports = CoreView.extend({
...
+    this.$('.js-header').prepend(this._searchView.render().el);
     this.addView(this._searchView);
   },
 
+  _renderActions: function () {
+    _.each(this.options.actions, function (a) {
+      var view = new CustomListAction(a);
+      this.$('.js-header').append(view.render().el);
+      this.addView(view);
+    }.bind(this));
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-view.js]: Good thing about _.each in underscore is you can pass the context as an argument (http://underscorejs.org/#each). No need to bind the scope to the function.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/multi-select-view.js [hunk 5]
```diff
@@ -96,11 +104,32 @@ Backbone.Form.editors.MultiSelect = Backbone.Form.editors.Base.extend({
       this._popupManager.untrack();
     }.bind(this);
 
-    this.collection.bind('change:selected', this._onItemSelected, this);
+    this.collection.bind('change:selected', _.debounce(this._onItemSelected, 50), this);
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/multi-select-view.js]: I encourage you to bind the this scope in this case.

File: lib/assets/core/test/spec/cartodb3/components/form-components/editors/select/multi-select-view.spec.js [hunk 6]
```diff
@@ -23,11 +24,80 @@ describe('components/form-components/editors/select/multi-select-view', function
     });
   });
 
-  it('should trigger change event on select item', function () {
+  it('should trigger change event on select item', function (done) {
     var onChange = jasmine.createSpy('onChange');
     this.view.on('change', onChange);
     this.view.collection.at(1).set({selected: true});
-    expect(onChange).toHaveBeenCalled();
+    setTimeout(function () {
```
Comment (xavijam) [inline lib/assets/core/test/spec/cartodb3/components/form-components/editors/select/multi-select-view.spec.js]: Better than use a timeout, https://jasmine.github.io/2.0/introduction.html#section-Mocking_the_JavaScript_Timeout_Functions, clocks!

File: lib/assets/core/test/spec/cartodb3/components/form-components/editors/select/multi-select-view.spec.js [hunk 7]
```diff
@@ -23,11 +24,80 @@ describe('components/form-components/editors/select/multi-select-view', function
...
+      done();
+    }, 100);
+  });
+
+  describe('all/none buttons', function () {
+    it('should trigger only one change event (debounce)', function (done) {
+      var onChange = jasmine.createSpy('onChange');
+      this.view.on('change', onChange);
+      this.view._selectAll();
+      setTimeout(function () {
```
Comment (xavijam) [inline lib/assets/core/test/spec/cartodb3/components/form-components/editors/select/multi-select-view.spec.js]: Same as above ^^^  🍄

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-view.js (hunk 4 above)
Comment (rjimenezda) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-view.js]: Good to know 🆒

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-search.tpl (hunk 2 above)
Comment (rjimenezda) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-list-search.tpl]: Sure, how about 'u-transparent' 'u-opaque'? @rubenmoya

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/multi-select-view.js (hunk 5 above)
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/select/multi-select-view.js]: Also, after talking with @rjimenezda, we could change the way we trigger that change, what about to make the change of the attribute silently and then trigger a custom event like 'allSelected' or 'noneSelected' at the end of the loop?

**Comment by saleiva** [comment]
I don't like the way the "cross" button appears... @urbanphes @piensaenpixel  any idea?

**Comment by saleiva** [comment]
a possible solution is to remove those ALL/NONE buttons when having a search query applied

**Comment by xavijam** [comment]
I like that @saleiva, but in any case, wouldn't you display the cross for "removing" the searched query?

**Comment by saleiva** [comment]
yes, but not at the same time than the ALL / NONE buttons

**Comment by xavijam** [comment]
I feel it, we are on the same page 💃 . @rubenmoya and @rjimenezda clear?

File: lib/assets/core/test/spec/cartodb3/components/form-components/editors/select/multi-select-view.spec.js (hunk 6 above)
Comment (rubenmoya) [inline lib/assets/core/test/spec/cartodb3/components/form-components/editors/select/multi-select-view.spec.js]: It seems clocks don't affect setTimeout functions deep inside the code (like _.debounce), so we went with a different approach, mocking _.debounce.

**Comment by xavijam** [comment]
Moving to development, but we are close!

File: lib/assets/core/javascripts/cartodb3/components/custom-list/custom-view.js [hunk 8]
```diff
@@ -113,11 +115,15 @@ module.exports = CoreView.extend({
   _renderActions: function () {
     _.each(this.options.actions, function (action) {
       var view = new CustomListAction(action);
-      this.$('.js-header').append(view.render().el);
+      this.$('.js-actions').append(view.render().el);
       this.addView(view);
     }, this);
   },
 
+  _setActionsVisibility: function () {
```
Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/components/custom-list/custom-view.js]: When we are listening to one key's changes `change:key`, the handler has the signature `function (model, value)`. This way we can save the `this.model.get('query')`:

```
_setActionsVisibility: function (model, query) {
   this.$('.js-actions').toggleClass('u-hide', query !== '');
}
```

**Comment by nobuti** [comment]
👍

**Comment by nobuti** [comment]
Frontend tests passing 🚀

**Pull request merged**